api/sms_provider/vonage.go

```diff
@@ -0,0 +1,78 @@
+package sms_provider
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/netlify/gotrue/conf"
+)
+
+const (
+	defaultVonageApiBase = "https://rest.nexmo.com"
+)
+
+type VonageProvider struct {
+	Config  *conf.VonageProviderConfiguration
+	APIPath string
+}
+
+type VonageResponseMessage struct {
+	Status    string `json:"status"`
+	ErrorText string `json:"error-text"`
+}
+
+type VonageResponse struct {
+	Messages []VonageResponseMessage `json:"messages"`
+}
+
+// Creates a SmsProvider with the Vonage Config
+func NewVonageProvider(config conf.VonageProviderConfiguration) (SmsProvider, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
+	}
+
+	apiPath := defaultVonageApiBase + "/sms/json"
+	return &VonageProvider{
+		Config:  &config,
+		APIPath: apiPath,
+	}, nil
+}
+
+// Send an SMS containing the OTP with Vonage's API
+func (t VonageProvider) SendSms(phone string, message string) error {
+	body := url.Values{
+		"from":       {t.Config.From},
+		"to":         {phone},
+		"text":       {message},
+		"api_key":    {t.Config.ApiKey},
+		"api_secret": {t.Config.ApiSecret},
+	}
+
+	client := &http.Client{}
+	r, err := http.NewRequest("POST", t.APIPath, strings.NewReader(body.Encode()))
+	if err != nil {
+		return err
+	}
+
+	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+	res, err := client.Do(r)
+	if err != nil {
+		return err
+	}
+	defer res.Body.Close()
+
+	resp := &VonageResponse{}
+	derr := json.NewDecoder(res.Body).Decode(resp)
+	if derr != nil {
+		return derr
+	}
+
+	if resp.Messages[0].Status != "0" {
```

Comment:
will `resp.Messages` ever be an empty slice? if so, this could get ugly and panic

Reply:
what does it mean when `Status="0"`? if it's a vonage api specific jargon, could you add a comment to describe it or add a link to the api reference please?

Reply:
Here is the Vonage API status code https://developer.vonage.com/messaging/sms/guides/troubleshooting-sms
The above page list all possible status code. I don't think `resp.Messages` will be empty, it will have at least one member. When you send an SMS, the SMS API returns a JSON response that contains an array of message objects, one for each message:

Reply:
A status of zero indicates success; a non-zero value means something went wrong.

Reply:
i see, i think it's still safer to check the length of the slice to ensure we're not accessing an empty slice here.

> A status of zero indicates success; a non-zero value means something went wrong.
Would be great if you include this in the line 72 for future reference! Reason for this is because status "0" isn't immediately intuitive for someone with no experience with the vonage api.

Reply:
@kangmingtay Sure. I have added those changes.